File: solutions/haiku-1/haiku/java/primes.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();

        if (n < 2) {
            System.out.println();
            return;
        }

        // Sieve of Eratosthenes
        boolean[] isPrime = new boolean[n + 1];
        for (int i = 2; i <= n; i++) {
            isPrime[i] = true;
        }

        for (int i = 2; i * i <= n; i++) {
            if (isPrime[i]) {
                for (int j = i * i; j <= n; j += i) {
                    isPrime[j] = false;
                }
            }
        }

        // Collect and output primes
        StringBuilder result = new StringBuilder();
        for (int i = 2; i <= n; i++) {
            if (isPrime[i]) {
                if (result.length() > 0) {
                    result.append(" ");
                }
                result.append(i);
            }
        }

        System.out.println(result.toString());
    }
}

File: solutions/haiku-1/haiku/javascript/primes.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false
});

rl.on('line', (line) => {
  const n = parseInt(line.trim());

  // Handle case where n < 2 (no primes)
  if (n < 2) {
    console.log('');
    rl.close();
    return;
  }

  // Sieve of Eratosthenes
  const isPrime = new Array(n + 1).fill(true);
  isPrime[0] = false;
  isPrime[1] = false;

  for (let i = 2; i * i <= n; i++) {
    if (isPrime[i]) {
      for (let j = i * i; j <= n; j += i) {
        isPrime[j] = false;
      }
    }
  }

  // Collect all primes
  const primes = [];
  for (let i = 2; i <= n; i++) {
    if (isPrime[i]) {
      primes.push(i);
    }
  }

  console.log(primes.join(' '));
  rl.close();
});
